nssp/delphi_nssp/run.py

```diff
@@ -0,0 +1,168 @@
+            "missing_se",
+            "missing_sample_size"]
+
+    for col_name in col_names:
+        df[col_name] = np.nan
+    df = add_default_nancodes(df)
+    return df
+
+
+def logging(start_time, run_stats, logger):
+    """Boilerplate making logs."""
+    elapsed_time_in_seconds = round(time.time() - start_time, 2)
+    min_max_date = run_stats and min(s[0] for s in run_stats)
+    csv_export_count = sum(s[-1] for s in run_stats)
+    max_lag_in_days = min_max_date and (datetime.now() - min_max_date).days
+    formatted_min_max_date = min_max_date and min_max_date.strftime("%Y-%m-%d")
+    logger.info(
+        "Completed indicator run",
+        elapsed_time_in_seconds=elapsed_time_in_seconds,
+        csv_export_count=csv_export_count,
+        max_lag_in_days=max_lag_in_days,
+        oldest_final_export_date=formatted_min_max_date,
+    )
+
+
+def run_module(params):
+    """
+    Run the indicator.
+
+    Arguments
+    --------
+    params:  Dict[str, Any]
+        Nested dictionary of parameters.
+    """
+    start_time = time.time()
+    logger = get_structured_logger(
+        __name__,
+        filename=params["common"].get("log_filename"),
+        log_exceptions=params["common"].get("log_exceptions", True),
+    )
+    export_dir = params["common"]["export_dir"]
+    socrata_token = params["indicator"]["socrata_token"]
+    # if "archive" in params:
+    #     daily_arch_diff = S3ArchiveDiffer(
+    #         params["archive"]["cache_dir"],
+    #         export_dir,
+    #         params["archive"]["bucket_name"],
+    #         "nchs_mortality",
+    #         params["archive"]["aws_credentials"],
+    #     )
+    #     daily_arch_diff.update_cache()
+
+    run_stats = []
+    ## build the base version of the signal at the most detailed geo level you can get.
+    ## compute stuff here or farm out to another function or file
+    df_pull = pull_nssp_data(socrata_token)
+    ## aggregate
+    geo_mapper = GeoMapper()
+    for signal_i, signal in enumerate(SIGNALS):
+        for geo in GEOS:
+            df = df_pull.copy()
+            df["val"] = df[signal]
+            logger.info("Generating signal and exporting to CSV", metric=signal)
+            if geo == "nation":
+                df = df[df["geography"] == "United States"]
+                df["geo_id"] = "us"
+            elif geo == "state":
+                df = df[(df['county'] == "All") & (df["geography"] != "United States")]
+                df["geo_id"] = df["geography"].apply(lambda x: us.states.lookup(x).abbr.lower() if us.states.lookup(x) else 'dc')
+            elif geo == "hrr" or geo == "msa":
+                df = df[['fips', 'val', 'timestamp']]
+                df = geo_mapper.add_population_column(df, geocode_type="fips", geocode_col="fips")
+                df = geo_mapper.add_geocode(df, "fips", geo, from_col="fips", new_col="geo_id")
+                df = generate_weights(df, "val")
+                df = weighted_geo_sum(df, "geo_id", "val")
```

Comment:
note: once the population and new geo_id columns have been added on, a simple weighted mean can be done with

```
df.groupby(["timestamp", "geo_id"]).apply(lambda x: sum(x.population * x.val) / x.population.sum() )
```

but this doesn't handle the `NaN`s appropriately. Actually I'm surprised that there isn't a built-in weighted mean function. `numpy` has an [average function that takes weights](https://numpy.org/doc/stable/reference/generated/numpy.average.html) but also doesn't let you ignore `NaN`s.

Reply:
yeah, I had to write these as a way to handle `NaN` correctly for nwss and they're getting reused here. I plan to eventually add them to the geomapper. It is slower than using precomputed weights, so I'm also debating adding those. There's probably also room for optimizing them though.

Reply:
Yeah, at least an acknowledgement of the assumption would be good to include in the documentation for this data source.

Reply:
Talking with @dshemetov, apparently in the HRR case, the weighted fips -> hrr map is already present, so we may not need to do this. He may add the weighted fips -> msa map in the near future, since the data is already there for it